File: apps/desktop/src-tauri/src/commands/app_settings.rs
// 应用级设置 Tauri Commands：app_settings.get / set / reset
// 透传 sidecar JSON-RPC，前端 ↔ 后端共用 camelCase 字段。
// 文件行数硬上限 300。

use crate::sidecar::manager::SidecarState;
use serde::Serialize;
use serde_json::{json, Value};
use tauri::State;

/// app.settings.get 返回结构
#[derive(Debug, Serialize, Clone)]
pub struct AppSettingsResponse {
    /// 当前设置（默认值已合并），字段为 camelCase JSON object
    pub settings: Value,
    /// 默认值（前端"重置"按钮可一键填回）
    pub defaults: Value,
    /// 实际持久化文件路径（展示给高级用户/排错）
    pub path: String,
}

/// 读取应用设置（含默认值合并）。
#[tauri::command]
pub async fn app_settings_get(
    sidecar: State<'_, SidecarState>,
) -> Result<AppSettingsResponse, String> {
    let mut manager = sidecar.lock().map_err(|e| format!("锁 sidecar 失败: {e}"))?;
    if !manager.is_running() {
        manager.start().map_err(|e| format!("启动 sidecar 失败: {e}"))?;
    }
    let result = manager.call("app.settings.get", json!({}))?;
    Ok(AppSettingsResponse {
        settings: result.get("settings").cloned().unwrap_or_else(|| json!({})),
        defaults: result.get("defaults").cloned().unwrap_or_else(|| json!({})),
        path: result
            .get("path")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
    })
}

/// 部分更新应用设置；patch 是 camelCase JSON object。
#[tauri::command]
pub async fn app_settings_set(
    sidecar: State<'_, SidecarState>,
    patch: Value,
) -> Result<AppSettingsResponse, String> {
    let mut manager = sidecar.lock().map_err(|e| format!("锁 sidecar 失败: {e}"))?;
    if !manager.is_running() {
        manager.start().map_err(|e| format!("启动 sidecar 失败: {e}"))?;
    }
    let result = manager.call("app.settings.set", json!({ "patch": patch }))?;
    Ok(AppSettingsResponse {
        settings: result.get("settings").cloned().unwrap_or_else(|| json!({})),
        defaults: result.get("defaults").cloned().unwrap_or_else(|| json!({})),
        path: result
            .get("path")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
    })
}

/// 重置应用设置为默认值。
#[tauri::command]
pub async fn app_settings_reset(
    sidecar: State<'_, SidecarState>,
) -> Result<AppSettingsResponse, String> {
    let mut manager = sidecar.lock().map_err(|e| format!("锁 sidecar 失败: {e}"))?;
    if !manager.is_running() {
        manager.start().map_err(|e| format!("启动 sidecar 失败: {e}"))?;
    }
    let result = manager.call("app.settings.reset", json!({}))?;
    Ok(AppSettingsResponse {
        settings: result.get("settings").cloned().unwrap_or_else(|| json!({})),
        defaults: result.get("defaults").cloned().unwrap_or_else(|| json!({})),
        path: result
            .get("path")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
    })
}

File: apps/desktop/src/ipc/app_settings.ts
// 应用级设置 IPC：app.settings.get / set / reset。
// 调用 Tauri Rust commands，后者透传到 sidecar JSON-RPC。
// 字段为 camelCase，与 sidecar 端 DEFAULT_SETTINGS 对齐。

import { invoke } from "@tauri-apps/api/core";

/**
 * 应用级设置（前端可编辑字段）。
 *
 * 新增字段流程：
 *   1. sidecar 的 app_settings.py DEFAULT_SETTINGS 加默认值
 *   2. 这里添加 TypeScript 字段
 *   3. routes/Settings.tsx 的 GeneralTab 加输入控件
 */
export interface AppSettings {
  /** 通用 nexus-tool 默认超时（秒）。manifest.implementation.timeout 可单工具覆盖。 */
  nexusToolDefaultTimeoutSec: number;
  /** 同时允许运行的通用 nexus-tool 数。 */
  nexusToolMaxConcurrent: number;
  /** cancel 时是否递归杀子进程（Windows 下尤其重要）。 */
  nexusToolKillProcessTree: boolean;
  /** sidecar 日志等级（仅展示，热更新需后续迭代）。 */
  logLevel: "DEBUG" | "INFO" | "WARN" | "ERROR";
  /** 运行工具前是否自动安装 manifest.dependencies 中缺失的 Python 包。 */
  nexusToolAutoInstallDeps: boolean;
  /** pip 安装镜像源 URL（为空时使用默认 PyPI）。 */
  nexusToolPipMirror: string;
}

export interface AppSettingsResponse {
  /** 当前设置（默认值已合并） */
  settings: AppSettings;
  /** 默认值（"重置"按钮可一键填回） */
  defaults: AppSettings;
  /** 持久化文件路径（展示给高级用户排错） */
  path: string;
}

/** 读取应用设置（含默认值合并） */
export async function getAppSettings(): Promise<AppSettingsResponse> {
  return invoke<AppSettingsResponse>("app_settings_get");
}

/** 部分更新应用设置；patch 只需包含要改的字段 */
export async function patchAppSettings(
  patch: Partial<AppSettings>,
): Promise<AppSettingsResponse> {
  return invoke<AppSettingsResponse>("app_settings_set", { patch });
}

/** 重置为默认值 */
export async function resetAppSettings(): Promise<AppSettingsResponse> {
  return invoke<AppSettingsResponse>("app_settings_reset");
}
